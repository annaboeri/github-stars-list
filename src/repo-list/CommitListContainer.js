import React, { Component } from 'react';
import Commit from './Commit';

class CommitListContainer extends Component {
  constructor(props) {
    super(props);
    this.state = {
      error: null,
      isLoaded: false,
      commits: [],
    };
    this.fetchCommits = this.fetchCommits.bind(this);
  }

  componentDidMount() {
    this.fetchCommits();
  }

  calcYesterdaysDate() {
    const today = new Date().getTime();
    const yesterday = new Date(today - (24 * 60 * 60 * 1000));
    return yesterday.toISOString();
  }

  fetchCommits() {
    const query = `since=${this.calcYesterdaysDate()}`
    const apiUrl = `https://api.github.com/repos/${this.props.repoName}/commits?${query}`;
    fetch(apiUrl)
      .then(this.props.handleErrors)
      .then(
        (result) => {
          this.setState({
            isLoaded: true,
            commits: result
            });
        },
        (error) => {
          this.setState({
            isLoaded: true,
            error
          });
        }
      )
  }

  render() {
    const { error, isLoaded, commits } = this.state;
    const numberOfCommits = commits.length;

    if (error) {
      return (
         <section className="error-container">
           <div className="fa-2x">
             <i className="far fa-frown"></i>
           </div>
           <p>Unable to load commits</p>
          </section>
       )
    }
    else if (!isLoaded) {
      return (
        <section className="loading-container">
          <i className="fas fa-spinner fa-pulse"></i>
          <p>Loading...</p>
        </section>
      )
    } else {
      return (
        <div className="commit-list-container">
        {this.state.isLoaded && numberOfCommits < 1 ? (
          <div className="empty-commit-list">
            <h3>No commits in the last 24 hrs</h3>
          </div>
        ) : (
          <div>
            <h3>
              {`${numberOfCommits}${' '}
              commit${numberOfCommits > 1 ? 's' : ''} in the last 24 hrs`}
            </h3>
            <ul>
            {commits.map(commit =>
              <Commit commit={commit} key={commit.sha} />
            )}
            </ul>
          </div>
        )}
        </div>
      );
    }
  }
}

export default CommitListContainer;
