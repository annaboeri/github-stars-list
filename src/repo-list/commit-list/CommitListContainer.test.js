import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import CommitListContainer from './CommitListContainer';

describe('CommitListContainer', () => {
  it('should render loading state', (done) => {
    const component = shallow(<CommitListContainer />);
    const tree = toJson(component);

    expect(tree).toMatchSnapshot();
    expect(component.find('.loading-container').exists()).toBe(true);
    done();
  });

  it('should render commits', (done) => {
    const component = shallow(<CommitListContainer />);
    const tree = toJson(component);
    const mockState = { isLoaded: true }
    component.setState({ ...mockState });

    expect(tree).toMatchSnapshot();
    expect(component.find('.commit-list-container').exists()).toBe(true);
    done();
  });

  it('should render error state', (done) => {
    const component = shallow(<CommitListContainer />);
    const tree = toJson(component);
    const mockState = { error: true }
    component.setState({ ...mockState });

    expect(tree).toMatchSnapshot();
    expect(component.find('.error-container').exists()).toBe(true);
    done();
  });
});
