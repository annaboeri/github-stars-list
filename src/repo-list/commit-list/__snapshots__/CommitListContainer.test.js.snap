// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommitListContainer should render commits 1`] = `
<section
  className="loading-container"
>
  <i
    className="fas fa-spinner fa-pulse"
  />
  <p>
    Loading...
  </p>
</section>
`;

exports[`CommitListContainer should render error state 1`] = `
<section
  className="loading-container"
>
  <i
    className="fas fa-spinner fa-pulse"
  />
  <p>
    Loading...
  </p>
</section>
`;

exports[`CommitListContainer should render loading state 1`] = `
<section
  className="loading-container"
>
  <i
    className="fas fa-spinner fa-pulse"
  />
  <p>
    Loading...
  </p>
</section>
`;
